src/ecies.cpp

```diff
@@ -0,0 +1,466 @@
+#include "mococrw/ecies.h"
+#include "mococrw/symmetric_crypto.h"
+#include "mococrw/openssl_wrap.h"
+
+namespace mococrw {
+
+class ECIESImpl
+{
+public:
+    ECIESImpl(const std::shared_ptr<KeyDerivationFunction> kdf,
+              const size_t macKeySize,
+              std::function<std::unique_ptr<MessageAuthenticationCode> (const std::vector<uint8_t> &)> macFactoryFunc,
+              const size_t symCipherKeySize,
+              std::function<std::unique_ptr<SymmetricCipherI>(const std::vector<uint8_t>&,
+                                                              const SymmetricCipherOperation&)> symCipherFactoryFunc,
+              const std::vector<uint8_t> &S1,
+              const std::vector<uint8_t> &S2)
+        : kdf(kdf), _macKeySize(macKeySize), _macFactoryFunc(macFactoryFunc),
+          _symCipherKeySize(symCipherKeySize), _symCipherFactoryFunc(symCipherFactoryFunc),
+          _S1(S1), _S2(S2) {}
+
+    virtual ~ECIESImpl()
+    {
+        /* Write zeros into _sharedSecret and both keys */
+        _key_encrypt.assign(_key_encrypt.size(), 0);
+        _key_mac.assign(_key_mac.size(), 0);
+        _sharedSecret.assign(_sharedSecret.size(), 0);
+    }
+
+protected:
+    void createSymmetricKeys()
+    {
+        std::vector<uint8_t> result = kdf->deriveKey(_sharedSecret, _macKeySize + _symCipherKeySize, _S1);
+
+        /* Copy first part (encryption key) */
+        std::vector<uint8_t>::const_iterator first = result.begin();
+        std::vector<uint8_t>::const_iterator last = result.begin() + _symCipherKeySize;
+        _key_encrypt = std::vector<uint8_t>(first, last);
```

Comment:
```suggestion
        std::copy(result.begin(), result.begin() + _symCipherKeySize, _key_encrypt.begin());
```

or even nicer maybe (would be one copy less)

```
void createSymmetricKeys()
{
        std::vector<uint8_t> result = kdf->deriveKey(_sharedSecret, _macKeySize + _symCipherKeySize, _S1);
        std::copy(result.begin() + _symCipherKeySize, result.end(), _key_mac.begin()
        // add wipeout for bytes that will be removed here.
        result.resize(_symCipherKeySize);
        _key_encrypt = std::move(result);
```

Reply:
max code example changes the semantics on how the key's are obtained from the KDF output and thus influence the overall algorithm. So, I guess this doesn't work here because it part of the specification from which part of the KDF output which key needs to come from.

Reply:
bought!

Reply:
Took the first approach.
Didn't find a solution to wipe the vector without a copy operation. So resizing doesn't work if you cannot wipe it before easily.